compiler/code-gen/code-generator.cpp

```diff
@@ -0,0 +1,84 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "compiler/compiler-core.h"
+#include "compiler/code-gen/code-generator.h"
+#include "compiler/data/src-file.h"
+
+
+void CodeGenerator::open_file_create_writer(bool compile_with_debug_info_flag, bool compile_with_crc, const std::string &file_name, const std::string &subdir) {
+  std::string full_file_name = G->cpp_dir;
+  if (!subdir.empty()) {
+    full_file_name += subdir;
+    full_file_name += "/";
+  }
+  full_file_name += file_name;
+
+  // get a currently opened file from a cpp index saved at previous kphp launch — to compare hashes with it
+  // (if it doesn't exist in index, its hashes will be -1; if !compile_with_crc, the same)
+  cur_file = G->get_file_info(std::move(full_file_name));
+  cur_file->compile_with_debug_info_flag = compile_with_debug_info_flag;
+  cur_file->needed = true;
+  // also, cur_file->includes will be modified on demand, see add_include()
+
+  if (is_step_just_calc_hashes) {
+
+  } else {
+    kphp_assert(data == nullptr);
+    data = new WriterData(cur_file, compile_with_crc);
+    data->begin_line();
+  }
```

Comment:
Can rewrite this condition with negation to get rid of the empty body?

Reply:
It's intentionally. See the same if/else below (in the opposite function). This emphasizes, that on a file open, if we are calculating hashes, we do nothing.

Reply:
But yeah, it's a good point. I'll add a comment inside this empty if

Reply:
Understand. It will look good with a comment.